code-editor/scripts/python-package/src/sagemaker_code_editor_auto_shut_down/autostop.py

```diff
@@ -0,0 +1,74 @@
+import os
+import time
+import boto3
+import json
+import argparse
+
+def check_user_activity(workspace_dir, idle_threshold):
+    # Get the timestamp of the most recently modified file or folder
+    recent_item = max(
+        (os.path.join(root, file) for root, _, files in os.walk(workspace_dir) for file in files),
+        key=lambda x: os.stat(x).st_mtime,
+        default=None
+    )
+
+    # Get the current time
+    current_time = time.time()
+
+    # Calculate the time difference
+    time_diff = current_time - os.stat(recent_item).st_mtime if recent_item else float('inf')
+
+    # Check if the user is idle based on the idle time threshold
+    if time_diff > idle_threshold:
+        return "idle"
+    else:
+        return "active"
+
+# Create an argument parser
+parser = argparse.ArgumentParser(description='Check user activity and terminate SageMaker Studio app if idle.')
+parser.add_argument('--time', type=int, help='Idle time threshold in seconds')
+parser.add_argument('--region', type=str, help='AWS region')
+
+# Parse the command-line arguments
+args = parser.parse_args()
+
+# Check if idle_threshold is provided
+if args.time is None:
+    parser.print_help()
+    sys.exit(1)
+
+if args.region is None:
+    parser.print_help()
+    sys.exit(1)
+
+# Monitor workspace_dirs for changes to implement auto-shutdown, as these paths track updates to both unsaved and saved editor content, covering all user activity scenarios.
+workspace_dirs = ["/opt/amazon/sagemaker/sagemaker-code-editor-server-data/data/User/History", "/opt/amazon/sagemaker/sagemaker-code-editor-server-data/data/User/Backups/empty-window/untitled"]
+idle_threshold = args.time  # this is in seconds. for ex: 1800 seconds for 30 minutes
+aws_region = args.region # get the region.
+
+# Track the activity status for each directory
+activity_status = [check_user_activity(directory, idle_threshold) for directory in workspace_dirs]
```

Comment:
Do we have a file (or dir) in this directory tree that gets modified when the Code Editor application is opened first in the browser? If not, we need to track some state.

Reply:
As far as I know we do not, but I will check in and get back to you on this

Reply:
@giuseppeporcelli, we do not track file change in those directories when Code Editor is not opened - what state do you recommend that we track here?

Reply:
resolved with 1:1 discussion with @aws-spenceng